### ISSUE TITLE
Syncing Error with Ankidroid Insufficient memory

### ISSUE THREAD

**Issue by hssm** [ankidroid/Anki-Android#2649]
Originally reported on Google Code with ID 1747

```
What steps will reproduce the problem?
1.Syncing through ankidroid
2.
3.

What is the expected output? What do you see instead?
Syncing with desktop to add new decks/new cards to ankidroid, then error message pops
up saying "operation not possible due to insufficient memory on your device" even though
I have confirmed I have sufficient memory.  


Does it happen again every time you repeat the steps above? Or did it
happen only one time?
Everytime

What version of AnkiDroid are you using? (Decks list > menu > About > Look
at the title)
On what version of Android? (Home screen > menu > About phone > Android
version)
Ankidroid v2.0.1
Android version 2.3.4

If it is a crash or "Force close" and you can reproduce it, the following
would help immensely: 1) Install the "SendLog" app, 2) Reproduce the crash,
3) Immediately after, launch SendLog, 4) Attach the resulting file to this
report. That will make the bug much easier to fix.


Please provide any additional information below.

```

Reported by `danhonda` on 2013-04-24 16:15:42

<hr>
- _Attachment: [SendLog[1].txt](https://storage.googleapis.com/google-code-attachments/ankidroid/issue-1747/comment-0/SendLog[1].txt)_

**Comment by mikehardy** [comment]
Cross linking to #4481  and #4864 - I am intending to look at online synchronization after I finish with package import

**Comment by mikehardy** [comment]
The code has moved but this is definitely still possible in the code if the stream received is large enough compared to resources:
https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/sync/HttpSyncer.java#L299
https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/sync/RemoteServer.java#L115
https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/sync/RemoteServer.java#L87
https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/sync/Syncer.java#L179

**Comment by mikehardy** [comment]
I also just asked on #3282 if there was a way to know from the server how much data would come down. I'm unsure if we can do partial synchronization but for very large syncs that may be a possibility as well - sync in stages

**Comment by amlynch** [comment]
Occurred for me on 2.9.7 during May 2020. Ankidroid and Anki were already syncing prior to issue. Added a large deck on desktop Anki, and error appeared during next Ankidroid sync. Was using external SD card as AnkiDroid sync location. Had more than enough space on external SD and on internal memory. Device is Samsung Tab A.

My workaround was manually moving media to external SD card via inserting into laptop. Error no longer appears on sync.

**Comment by mikehardy** [comment]
Interesting @amlynch - that makes me think that the media files might have been the problem with the sync. I've not ever tested to see if (for whatever reason) a very large media file would do it. I've synced decks with 20,000+ images in them and that worked though. Is there any particular piece of media in this deck that is particularly large? Just as a guess, low chance of this being it but worth a check. If you're not super unix savvy I usually do something like `du -sk * | sort -n` in order to find the biggest file in a directory, it'll be the last line, in kilobytes

**Comment by amlynch** [comment]
@mikehardy, the deck I was syncing is about 30,000 cards, with 21,207 images. In the collection.media folder I see the biggest file as 5.5 MB; in fact only 200 files cross the 1 MB threshold.

*I'm using the size indicated in Finder; du gave me an "argument list too long" error, and I couldn't figure out an alternate. Let me know if du would reveal a different size than finder.

**Comment by mikehardy** [comment]
That seems like sufficient information on the collection, thank you @amlynch - I'm going to say based on that my hypothesis of a single large file is most likely wrong. You mention you are on an external SDcard which has historically caused us all sorts of problems. Can you tell me what filesystem the SDCard is using, and what device model you are on (manufacturer + model number). I wonder if it's small memory or the SDCard is still FAT and ran out of ability to make directory entries or something.

That last bit (ran out of directory entries) sounds implausible since a copy worked for you but I think during sync we might have multiple copies of each file sitting around temporarily leading to a situation where we need 2x the entries available or similar.

I'd be curious to know if you could reproduce it with internal storage but that may not be feasible if it doesn't fit

**Comment by amlynch** [comment]
Sure thing; sorry for the initial lack of detail. I think I reformatted SD from FAT to ExFAT (unfortunately I'm not 100% sure it wasn't in ExFAT initially, neglected to keep good notes).

**Laptop**: MacBook Air, Catalina 10.15.4
**Tablet**: Samsung Tab A 2019 (SM-T290)
**SD Card**: Samsung EVO Select microSD Memory Card 32GB, ExFAT (May have reformatted via Disk Utility)

I created a com.ichi2.anki directory on the SD card, then redirected AnkiDroid to look there. Initial AnkiWeb sync of small deck was successful. Added larger deck to desktop app, then synced again; AnkiDroid error message occurred (“operation not possible due to insufficient memory on your device”). Clearing the error message returned me to AnkiDroid. All of the _cards_ from the new deck downloaded, but only some of the _media files_.

 I then inserted SD card into laptop, (possibly) reformatted as ExFAT, created the appropriate directories, and used rsync to move files from my local collection.media to the SD’s collection.media.  AnkiDroid is now running fine and media is successfully loading on cards. 

So the thought about FAT being an issue might be accurate. You’re correct I couldn’t reproduce on just internal as there isn’t enough memory to get to the stage with that error message.
